File: backend/catalogService/src/main/java/levin/schnabel/catalogService/structural/entity/Category.java
package levin.schnabel.catalogService.structural.entity;

import java.util.Arrays;
import java.util.Set;
import java.util.stream.Collectors;

public enum Category {

	SPORTS(null),
	COMPETITIVE(SPORTS),
	VOLLEYBALL(COMPETITIVE),
	BEACH_VOLLEYBALL(VOLLEYBALL),
	INDOOR_VOLLEYBALL(VOLLEYBALL),
	TABLE_TENNIS(COMPETITIVE),
	TABLE_TENNIS_RACKETS(TABLE_TENNIS),
	TABLE_TENNIS_EQUIPMENT(TABLE_TENNIS),

	CASUAL(SPORTS),
	ROPE_SKIPPING(CASUAL),
	RUNNING(CASUAL),
	SHOES(RUNNING),
	RUNNING_TRACKERS(RUNNING),
	CYCLING(CASUAL),
	BIKES(CYCLING),
	BIKE_COMPONENTS(CYCLING),
	HELMETS(CYCLING);


	private final Category parent;

	Category(Category parent) {
		this.parent = parent;
	}

	public Category getParent() {
		return parent;
	}

	public String getName() {
		return this.name();
	}

	public Set<Category> getChildren() {
		return Arrays.stream(Category.values()).filter(category -> category.parent == this).collect(Collectors.toSet());
	}

	public Set<Category> getSiblings() {
		return Arrays.stream(Category.values()).filter(category -> category != this && category.parent == this.parent).collect(Collectors.toSet());
	}

}

File: frontend/SportsVendor/src/shared/model/Category.ts
export interface Category {
  displayName: string;
  name: string;
  // children: Category[],
  // parent: Category;
}

export const BEACH_VOLLEYBALL = {
  displayName: 'Beach Volleyball',
  name: 'BEACH_VOLLEYBALL'
}

export const INDOOR_VOLLEYBALL = {
  displayName: 'Indoor Volleyball',
  name: 'INDOOR_VOLLEYBALL'
}

export const TENNIS = {
  displayName: 'Tennis',
  name: 'TENNIS'
}

export const VOLLEYBALL = {
  displayName: 'Volleyball',
  name: 'VOLLEYBALL',
  children: [BEACH_VOLLEYBALL, INDOOR_VOLLEYBALL]
}

export const FOOTBALL = {
  displayName: 'Football',
  name: 'FOOTBALL'
}

export const CATEGORIES: Category[] = [
  VOLLEYBALL,
  BEACH_VOLLEYBALL,
  INDOOR_VOLLEYBALL,
  TENNIS,
  FOOTBALL
]
